_posts/en/2017-03-13-performance-optimizations-1.md

```diff
@@ -0,0 +1,117 @@
+---
+type: posts
+layout: post
+lang: en
+name: performance-optimizations-1
+id: en-performance-optimizations-1
+title: On-chain scaling: a review of historical performance optimization made to Bitcoin’s reference software. Part 1
+permalink: /en/2017/03/13/performance-optimizations-1/
+version: 1
+excerpt: Development milestones that helped preserve a reliable experience for users of the Bitcoin software client over the years. 
+---
+
+The following post aims to highlight development milestones that helped preserve a reliable experience for users of the Bitcoin software client over the years. We present several upgrades that were critical in maintaining the decentralized properties of the network and mitigate the resources burden of its participants. We describe how numerous orders-of-magnitude optimizations were made so that the Bitcoin network could support the growth in transaction activity without dramatically increasing the costs of participation for new and existing users. Finally, we note how those improvements all fall within a larger, systematic approach to protocol development that uses insights from Big-O complexity concepts and leverages smarter algorithms that make more efficient use of the network’s resources. 
+
+## Signature Caching
+Release: Bitcoin-Qt 0.7.0
+
+Verification of ECSDA signatures are one of the most computationally heavy operations done at the peer level. Because they need to be verified for every transactions, any superfluous validation leads to significant resource overhead for the end user. That was the case for early versions of the software which would both verify signatures before they entered a node mempool and after they were received into a block. 
+
+In order to gain in efficiency, developers created a cache allowing nodes to store previously validated signatures and skip redundant work once the transactions make it into an accepted block. 
+
+Additionally, the signature cache also mitigates a DoS vector introduced by the potential for specifically crafted transactions to stall Bitcoin clients. 
+
+### Further information
+
+ * [Bitcoin-Qt 0.7.0 Release notes](https://bitcoin.org/en/release/v0.7.0#core-bitcoin-handling-and-blockchain-database)
+ * [Fixed vulnerability explanation: Why the signature cache is a DoS protection](https://bitcointalk.org/index.php?topic=136422.0)
+
+## Ultraprune + LevelDB
+Release: Bitcoin-Qt 0.8.0
+
+Ultraprune was one of the first major upgrade to the Bitcoin software aimed at solving the overhead associated with validating transaction data from the blockchain. Previous versions of the Bitcoin reference client would use the entire transaction index which, even back then, resulted in a considerably heavy set of information to work from. One consequence of this was an almost unbearably slow initial block download, which turned out to be a significant hindrance for users looking to bootstrap a new client. 
```

Comment:
Let's reword this to say something like "Previous versions of the Bitcoin reference client maintained an index of all transaction outputs, spent or unspent. Ultraprune significantly reduced the size of that index with the insight that you only needed to keep track of unspent outputs, and an output - once spent - can be removed from the indexes entirely."